server/server.go

```diff
@@ -1656,6 +1665,106 @@ func (s *StanServer) start(runningState State) error {
 	return nil
 }
 
+// startChannelGossiping sets up a NATS subscription to a cluster gossip
+// subject to receive messages from the cluster on what channels exist. It also
+// starts a goroutine which periodically publishes the channels this server
+// knows about. When a server discovers it's missing channels, it requests the
+// missing channels from another server and creates them. This should only be
+// called if the server is running in clustered mode.
+func (s *StanServer) startChannelGossiping() error {
+	// Setup sub for receiving channel gossip.
+	gossipInbox := fmt.Sprintf("%s.%s.channels", defaultGossipPrefix, s.opts.ID)
+	_, err := s.ncr.Subscribe(gossipInbox, func(m *nats.Msg) {
+		c := &spb.ChannelGossip{}
+		if err := c.Unmarshal(m.Data); err != nil {
+			s.log.Errorf("Received invalid channel gossip message: %v", err)
+			return
+		}
+		if c.NodeID == s.opts.Clustering.NodeID {
+			// Ignore messages from ourselves.
+			return
+		}
+		s.channels.RLock()
+		count := uint32(len(s.channels.channels))
+		s.channels.RUnlock()
+
+		if count >= c.NumChannels {
+			return
+		}
+
+		// If our count is less, request the list of channels from the other
+		// server and reconcile.
+		resp, err := s.ncr.Request(m.Reply, nil, 5*time.Second)
+		if err != nil {
+			s.log.Errorf("Failed to fetch channels from another server: %v", err)
+			return
+		}
+		channelResp := &spb.ChannelResponse{}
+		if err := channelResp.Unmarshal(resp.Data); err != nil {
+			s.log.Errorf("Received invalid channels response from server: %v", err)
+			return
+		}
+		// Reconcile list of channels.
+		for _, channel := range channelResp.Channels {
+			if s.channels.get(channel) == nil {
+				if _, err := s.channels.createChannel(s, channel); err != nil {
+					s.log.Errorf("Failed to create channel %s while reconciling channels: %v", channel, err)
+				} else {
+					s.log.Debugf("Created channel %s while reconciling channels", channel)
+				}
+			}
+		}
+	})
+	if err != nil {
+		return err
+	}
+
+	// Setup sub for receiving channel requests.
+	reqInbox := fmt.Sprintf("%s.%s.channels.%s", defaultGossipPrefix, s.opts.ID, s.opts.Clustering.NodeID)
+	_, err = s.ncr.Subscribe(reqInbox, func(m *nats.Msg) {
+		s.channels.RLock()
+		channels := make([]string, len(s.channels.channels))
+		i := 0
+		for channel, _ := range s.channels.channels {
+			channels[i] = channel
+			i++
+		}
+		s.channels.RUnlock()
+		resp, err := (&spb.ChannelResponse{Channels: channels}).Marshal()
+		if err != nil {
+			panic(err)
+		}
+		s.ncr.Publish(m.Reply, resp)
+	})
+	if err != nil {
+		return err
+	}
+
+	// Start publishing channel info.
+	go func() {
```

Comment:
A bit late saying that, but with clustering we started to use quite a bit of go routines, and I would have like to make sure that we wait on them when server shutdown (more for tests than anything). In this case, you should call s.wg.Add(1) before the `go func()` and start the func with a `defer s.wg.Done()`.
When you have situations where you want to start a server go routine, and you are not currently holding the `s.mu.Lock()`, then you can use `StanServer.startGoRoutine()` which ensures that the server is not being shutdown, and if so, bumps the `s.wg` count and starts the given go routine.

Reply:
This is what the `shutdownCh` is for though? Or are you saying it's not sufficient to have the goroutine eventually exit but instead have shutdown only return once we know the goroutine has exited?

Reply:
Yes, the go routine will eventually stop, but sometimes, and again essentially for tests, it is better to ensure that things are done before moving to the next test, so yes, the idea is to have Shutdown() not return until go routines are done. It may not be an issue with this go routine though.

Reply:
.. but good practice as we start go routines.

Reply:
This is not a blocker, though. The splitting of the list to fit into NATS messages is.